### PR TITLE
ET debugger V2 (0.13.1)

### PR DESCRIPTION
**Pull request by improbable-danny**
#### Description
ET debugger work required for filtering APIs + rotating log APIs.

#### Tests
Unit tests + NFRs

#### PRs
- https://github.com/improbable/TestGymBuildkite/pull/268
- https://github.com/spatialos/UnrealGDKTestGyms/pull/401/files
- https://github.com/improbable/nfr-benchmark-pipeline/pull/171

#### Primary reviewers
@MatthewSandfordImprobable

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 1]
```diff
@@ -273,16 +273,26 @@ FLocalDeploymentManager::ERuntimeStartResponse FLocalDeploymentManager::StartLoc
 	// Use the runtime start timestamp as the log directory, e.g. `<Project>/spatial/localdeployment/<timestamp>/`
 	FString LocalDeploymentLogsDir = FPaths::Combine(SpatialGDKServicesConstants::LocalDeploymentLogsDir, RuntimeStartTime.ToString());
 
+	// Store these logs alongside the GDK ones for convenience
+	const TCHAR* RuntimeEventLogPaths = TEXT("EventTracing/runtime");
+	FString EventTracingPath = FPaths::ConvertRelativePathToFull(FPaths::Combine(FPaths::ProjectSavedDir(), RuntimeEventLogPaths));
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	const FString FileName = TEXT("gdk");
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:280]: Do we use these?

File: SetupIncTraceLibs.bat [hunk 2]
```diff
@@ -18,8 +18,8 @@ call :MarkStartOfBlock "Create folders"
 call :MarkEndOfBlock "Create folders"
 
 call :MarkStartOfBlock "Retrieve dependencies"	
-	spatial package retrieve internal        trace-dynamic-x86_64-vc141_md-win32        %PINNED_CORE_SDK_VERSION% 	"%CORE_SDK_DIR%\trace_lib\trace-win32.zip"
```
Comment (MatthewSandfordImprobable) [inline SetupIncTraceLibs.bat:21]: Huh wonder why we did this?

Comment (MatthewSandfordImprobable) [reply SetupIncTraceLibs.bat:21]: Ahh ok this is explicitly done for this PR. Coool. wonder if we need to update the other references to the runtime version? There are some places referencing 15.1.0 -preview I believe?

File: SpatialGDK/Source/SpatialGDKServices/Public/SpatialGDKServicesConstants.h [hunk 3]
```diff
@@ -52,7 +52,7 @@ static const FString GetInspectorExecutablePath(const FString& InspectorVersion)
 	return FPaths::Combine(GDKProgramPath, InspectorPackageName, InspectorVersion, InspectorExe);
 }
 
-const FString SpatialOSRuntimePinnedStandardVersion = TEXT("15.0.1");
+const FString SpatialOSRuntimePinnedStandardVersion = TEXT("15.1.0-preview-1");
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Public/SpatialGDKServicesConstants.h:55]: perhaps we revert this?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/WorkerVersionCheck.h [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 #include "improbable/c_worker.h"
 
-#define WORKER_SDK_VERSION "15.0.1"
+#define WORKER_SDK_VERSION "15.1.0"
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/WorkerVersionCheck.h:7]: perhaps we revert this?

File: SpatialGDK/Extras/core-sdk.version [hunk 5]
```diff
@@ -1,2 +1,2 @@
-15.0.1
+15.1.0-preview-1
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Extras/core-sdk.version:1]: perhaps we revert this?

File: SetupIncTraceLibs.bat (hunk 2 above)
Comment (improbable-danny) [reply SetupIncTraceLibs.bat:21]: 15.1.0 has fixed a compatibility problem with the legacy latency tracer (we can remove our hacks) - seemed too much for this PR though

File: SpatialGDK/Extras/core-sdk.version (hunk 5 above)
Comment (improbable-danny) [reply SpatialGDK/Extras/core-sdk.version:1]: Will update with 15.1.0 when the full release is out

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKDefaultLaunchConfigGenerator.cpp [hunk 6]
```diff
@@ -178,6 +178,39 @@ bool GenerateLaunchConfig(const FString& LaunchConfigPath, const FSpatialLaunchC
...
+			Writer->WriteObjectStart(TEXT("event_filter_configuration"));
+
+			UEventTracingSamplingSettings* SamplingSettings =
+				SpatialGDKSettings->GetEventTracingSamplingSettings(); // TODO: Is this trouble in editor?
+			Writer->WriteValue(TEXT("event_pre_filter"),
+							   SamplingSettings->RuntimeEventPreFilter.Len() ? *SamplingSettings->RuntimeEventPreFilter : TEXT("false"));
+			Writer->WriteValue(TEXT("event_post_filter"),
+							   SamplingSettings->RuntimeEventPostFilter.Len() ? SamplingSettings->RuntimeEventPostFilter : TEXT("false"));
+
+			Writer->WriteObjectEnd();
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKDefaultLaunchConfigGenerator.cpp:210]: minor: I like to add comments that tell you what object you are finishing when writing out markup blocks.
`Writer->WriteObjectEnd(); // event_tracing_configuration`

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 7]
```diff
@@ -273,16 +273,24 @@ FLocalDeploymentManager::ERuntimeStartResponse FLocalDeploymentManager::StartLoc
 	// Use the runtime start timestamp as the log directory, e.g. `<Project>/spatial/localdeployment/<timestamp>/`
 	FString LocalDeploymentLogsDir = FPaths::Combine(SpatialGDKServicesConstants::LocalDeploymentLogsDir, RuntimeStartTime.ToString());
 
+	// Store these logs alongside the GDK ones for convenience
+	const TCHAR* RuntimeEventLogPaths = TEXT("EventTracing/runtime");
+	FString EventTracingPath = FPaths::ConvertRelativePathToFull(FPaths::Combine(FPaths::ProjectSavedDir(), RuntimeEventLogPaths));
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	if (!PlatformFile.CreateDirectoryTree(*EventTracingPath))
+	{
+		UE_LOG(LogSpatialDeploymentManager, Log, TEXT("Failed to create runtime event log path."));
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:282]: Is this really a log category log? not a warning or an error?

**Review by simonsarginson — APPROVED** [review]
LGTM, don't have the most context but seems ok.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 7 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp]: Upgraded to error, should always be writable!

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 8]
```diff
@@ -273,16 +273,24 @@ FLocalDeploymentManager::ERuntimeStartResponse FLocalDeploymentManager::StartLoc
 	// Use the runtime start timestamp as the log directory, e.g. `<Project>/spatial/localdeployment/<timestamp>/`
 	FString LocalDeploymentLogsDir = FPaths::Combine(SpatialGDKServicesConstants::LocalDeploymentLogsDir, RuntimeStartTime.ToString());
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:274]: So we don't use this anymore. Are all all logs going to be put into the "EventTracing/runtime" folder?

Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:274]: Guess this is redundant now?

**Review by MatthewSandfordImprobable — APPROVED** [review]

File: SetupIncTraceLibs.bat (hunk 2 above)
Comment (m-samiec) [reply SetupIncTraceLibs.bat:21]: Does 15.1.0 have breaking api changes?

Comment (improbable-danny) [reply SetupIncTraceLibs.bat:21]: Yeah, the renaming of the legacy trace APIs is there :tada:

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp [hunk 9]
```diff
@@ -23,27 +40,34 @@ void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
 	}
 
 	uint32_t ItemSize = Trace_GetSerializedItemSize(Item);
-	if (EventTracer->BytesWrittenToStream + ItemSize <= EventTracer->MaxFileSize)
+	// Depends whether we are using rotating logs or single-log mode (where we track max size).
+	const bool bTrackFileSize = EventTracer->MaxFileSize != 0;
+	if (!bTrackFileSize || (EventTracer->BytesWrittenToStream + ItemSize <= EventTracer->MaxFileSize))
 	{
-		EventTracer->BytesWrittenToStream += ItemSize;
+		if (bTrackFileSize) // Tracked file size
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp:47]:
```suggestion
		if (bTrackFileSize) 
```

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 10]
```diff
@@ -94,8 +95,41 @@ void CheckCmdLineOverrideOptionalStringWithCallback(const TCHAR* CommandLine, co
 	UE_LOG(LogSpatialGDKSettings, Log, TEXT("%s is %s."), PrettyName,
 		   OverrideValue.IsSet() ? *(OverrideValue.GetValue()) : TEXT("not set"));
 }
+
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp:98]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.cxx [hunk 11]
```diff
@@ -205,6 +205,7 @@ const TArray<FString> ServerAuthorityWellKnownSchemaImports = {
     "unreal/gdk/unreal_metadata.schema",
     "unreal/gdk/actor_group_member.schema",
     "unreal/gdk/actor_set_member.schema",
+    "unreal/gdk/migration_diagnostic.schema",
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.cxx:208]: what's this doing here?

File: SetupIncTraceLibs.bat (hunk 2 above)
Comment (m-samiec) [reply SetupIncTraceLibs.bat:21]: Nice - we should make sure to revert this when we actually make the fix

**Review by m-samiec — APPROVED** [review]

**Pull request merged**